unix: fix assertion failure in linux-inotify.c; add a corresponding test

ahippo opened the pull request:
@saghul asked to post these changes separately from #620 (and #577 before), so that they can be included into 1.8.0 release without other larger changes introduced in those pull requests.

After QUEUE_MOVE() change in 442b8a5a848e1589520a4d4fd175d7e9aa084c44 linux-inotify.c started to assert under certain conditions.
Revert QUEUE_MOVE() change for linux-inotify.c and apply QUEUE_FOREACH_SAFE()/QUEUE_REMOVE_SAFE() approach to it instead.
Add a corresponding test-case (this is a bigger test-case that also checks that QUEUE_MOVE() works fine for other handle types).

Note, that GitHub shows commits in a wrong order (ordered by commit date).
The actual order is the following: cd623378ea0d534dac1d4383a6c11284ba5404e1 (test), c09cc1a9af6fa66c01a482b6c9193affe3ddd91a (revert), d9975c5959c6850dabf195f036d9a02259812dbd (new macros), 41645a80311972a3b06394966c242b985ef52d5d (linux-inotify.c fix).

saghul commented on test/test-queue-foreach-delete.c:
```diff
@@ -0,0 +1,211 @@
+
+#ifdef _WIN32
+static int mytouch(const char *filename) {
+  return _utime(filename, NULL);
```
why not use `uv_fs_utime`?

ahippo replied:
> why not use uv_fs_utime?

Thank you for the suggestion!
I think I decided to go with synchronous utime() call for simplicity.
I'll try to use uv_fs_utime().

saghul replied:
You can ue the fs functions synchronously by passing NULL as the callback.

ahippo replied:
> You can ue the fs functions synchronously by passing NULL as the callback.

Oh, wow, this helps a lot!
I didn't notice this feature.

saghul commented on test/test-queue-foreach-delete.c:
```diff
@@ -0,0 +1,211 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Make it "The libuv project and contributors"

ahippo replied:
> Make it "The libuv project and contributors"

Sure.

saghul commented on src/unix/linux-inotify.c:
```diff
@@ -35,6 +35,7 @@
 struct watcher_list {
   RB_ENTRY(watcher_list) entry;
   QUEUE watchers;
+  QUEUE* watchers_read_iter;
```
I wonder if this won't the ABI of any "public" structure.

saghul replied:
Ah, scrap that, this is defined and used just in this file.

saghul commented:
> Note, that GitHub shows commits in a wrong order (ordered by commit date).

You can reorder them as follows:

git rebase -i HEAD~4
(order them in your editor and save)
git push --force

saghul commented:
A quick comment: how about using a boolean value in watcher_list indicating it's being iterated? Then if so, uv_fsevent_stop would not do anything and we'd do the cleanup after calling all the callbacks for a given path watcher.

This should simplify the code, and work in all cases, since the safe macros don't work in 100% of the cases, according to the comment.

ahippo commented:
>> Note, that GitHub shows commits in a wrong order (ordered by commit date).

> You can reorder them as follows:
> 
> git rebase -i HEAD~4
> (order them in your editor and save)
> git push --force

Thank you for the hint!
This is GitHub-only problem, because it sorts commits by date.
(and dates here are all off because it was rebased and cherry-picked 100 times already)
The actual order in git (i.e. parent information) is already correct.

ahippo commented:
> A quick comment: how about using a boolean value in watcher_list indicating it's being iterated? Then if so, uv_fsevent_stop would not do anything and we'd do the cleanup after calling all the callbacks for a given path watcher.
> 
> This should simplify the code, and work in all cases, since the safe macros don't work in 100% of the cases, according to the comment.

Let me see, maybe this could work.
What do you mean by "safe macros don't work in 100% of the cases"?
Comments only state that you can't delete QUEUE head but I don't think we ever do so.

ahippo commented:
I've addressed all comments except for the last one ("boolean value in watcher_list") and updated the pull request.
I'll take a look at the last comment now.

ahippo commented:
>> A quick comment: how about using a boolean value in watcher_list indicating it's being iterated? Then if so, uv_fsevent_stop would not do anything and we'd do the cleanup after calling all the callbacks for a given path watcher.
>> 
>> This should simplify the code, and work in all cases, since the safe macros don't work in 100% of the cases, according to the comment.

> Let me see, maybe this could work.
> What do you mean by "safe macros don't work in 100% of the cases"?
> Comments only state that you can't delete QUEUE head but I don't think we ever do so.

Your approach also works.
The thing I don't like about it is that we use two tricks (QUEUE_MOVE() + this boolean flag) at the same time to make uv_fs_event_stop() work safely within uv__inotify_read().
QUEUE_*_SAFE() approach will do it in one shot.
But if you'd like, I can go ahead and update the pull request.

Another reason I chose QUEUE_*_SAFE() macros is that Win backend open-codes the same thing in [win/loop-watcher.c](../blob/v1.x/src/win/loop-watcher.c) (which can be converted to use these macros), and it's common to have *_safe() macros alongside regular ones for iterating over lists/queues (Linux's list.h, for example).


This is the draft changeset for your idea:
```diff
diff --git a/src/unix/linux-inotify.c b/src/unix/linux-inotify.c
index e3ff6d5..106ecf5 100644
--- a/src/unix/linux-inotify.c
+++ b/src/unix/linux-inotify.c
@@ -35,6 +35,7 @@
 struct watcher_list {
   RB_ENTRY(watcher_list) entry;
   QUEUE watchers;
+  int watchers_being_iterated_over;
   char* path;
   int wd;
 };
@@ -113,6 +114,14 @@ static struct watcher_list* find_watcher(uv_loop_t* loop, int wd) {
   return RB_FIND(watcher_root, CAST(&loop->inotify_watchers), &w);
 }
 
+static void maybe_free_watcher_list(struct watcher_list* w, uv_loop_t* loop) {
+  if ((!w->watchers_being_iterated_over) && QUEUE_EMPTY(&w->watchers)) {
+    /* No watchers left for this path. Clean up. */
+    RB_REMOVE(watcher_root, CAST(&loop->inotify_watchers), w);
+    uv__inotify_rm_watch(loop->inotify_fd, w->wd);
+    uv__free(w);
+  }
+}
 
 static void uv__inotify_read(uv_loop_t* loop,
                              uv__io_t* dummy,
@@ -161,15 +170,18 @@ static void uv__inotify_read(uv_loop_t* loop,
       path = e->len ? (const char*) (e + 1) : uv__basename_r(w->path);
 
       QUEUE_MOVE(&w->watchers, &queue);
+      w->watchers_being_iterated_over = 1;
       while (!QUEUE_EMPTY(&queue)) {
         q = QUEUE_HEAD(&queue);
         h = QUEUE_DATA(q, uv_fs_event_t, watchers);
 
         QUEUE_REMOVE(q);
         QUEUE_INSERT_TAIL(&w->watchers, q);
 
         h->cb(h, path, events, 0);
       }
+      w->watchers_being_iterated_over = 0;
+      maybe_free_watcher_list(w, loop);
     }
   }
 }
@@ -221,6 +233,7 @@ int uv_fs_event_start(uv_fs_event_t* handle,
   w->wd = wd;
   w->path = strcpy((char*)(w + 1), path);
   QUEUE_INIT(&w->watchers);
+  w->watchers_being_iterated_over = 0;
   RB_INSERT(watcher_root, CAST(&handle->loop->inotify_watchers), w);
 
 no_insert:
@@ -248,12 +261,7 @@ int uv_fs_event_stop(uv_fs_event_t* handle) {
   uv__handle_stop(handle);
   QUEUE_REMOVE(&handle->watchers);
 
-  if (QUEUE_EMPTY(&w->watchers)) {
-    /* No watchers left for this path. Clean up. */
-    RB_REMOVE(watcher_root, CAST(&handle->loop->inotify_watchers), w);
-    uv__inotify_rm_watch(handle->loop->inotify_fd, w->wd);
-    uv__free(w);
-  }
+  maybe_free_watcher_list(w, handle->loop);
 
   return 0;
 }

```

saghul commented:
> The thing I don't like about it is that we use two tricks (QUEUE_MOVE() + this boolean flag) at the same time to make uv_fs_event_stop() work safely within uv__inotify_read().
> QUEUE_*_SAFE() approach will do it in one shot.
> But if you'd like, I can go ahead and update the pull request.
> 
> Another reason I chose QUEUE_*_SAFE() macros is that Win backend open-codes the same thing in win/loop-watcher.c (which can be converted to use these macros), and it's common to have *_safe() macros alongside regular ones for iterating over lists/queues (Linux's list.h, for example).

I don't disagree, but the _safe variants are a bigger change than I'd like to handle right now, so I'd rather stay with a simpler fix, and then introduce the _safe variants as a refactor in the entire codebase, if need be.

Your fix is exactly what I had in mind :-) can you incorporate it? (please find a shorter name for "watchers_being_iterated_over", maybe "iterating"?)

ahippo commented:
> I don't disagree, but the _safe variants are a bigger change than I'd like to handle right now, so I'd rather stay with a simpler fix, and then introduce the _safe variants as a refactor in the entire codebase, if need be.

Ok, makes sense.

> Your fix is exactly what I had in mind :-) can you incorporate it? (please find a shorter name for "watchers_being_iterated_over", maybe "iterating"?)

Great!
Sure, I run out of inspiration yesterday.:)

ahippo commented:
@saghul, I've updated the pull request and incorporated the change.

saghul commented:
CI: https://ci.nodejs.org/job/libuv+any-pr+multi/187/

saghul commented:
Test fails on OSX:

~~~~
[%  99|+ 278|-   1|T   0|S   0]: queue_foreach_delete
`queue_foreach_delete` failed: timeout
Output from process `queue_foreach_delete`:
Assertion failed in ../test/test-queue-foreach-delete.c on line 201: fs_event_cb_calls[0] == 1
=============================================================
~~~~

SmartOS:

~~~~
[%  99|+ 271|-   3|T   0|S   3]: queue_foreach_delete
`queue_foreach_delete` failed: exit code 134
Output from process `queue_foreach_delete`:
Assertion failed in test/test-queue-foreach-delete.c on line 201: fs_event_cb_calls[2] == 1
=============================================================
~~~~

and Windows:

~~~~
[%  99|+ 269|-   0|T   0|S   3]: queue_foreach_delete
`queue_foreach_delete` failed: exit code 3
Output from process `queue_foreach_delete`:
Assertion failed in test\test-queue-foreach-delete.c on line 200: async_cb_calls[0] == 1
=============================================================
~~~~

ahippo commented:
> Test fails on OSX:
> SmartOS:
> and Windows:

And FreeBSD as far as I can see.

Hmm, I'll take a look.
Most likely the order in which callbacks are called is backend-dependent, I likely need to make checks backend-dependent too.

ahippo commented:
> Hmm, I'll take a look.
> Most likely the order in which callbacks are called is backend-dependent, I likely need to make checks backend-dependent too.

I don't have much time to investigate the problem now.:(
So, let's just RETURN_SKIP() on all platforms except Linux for now.
The main purpose of the test is to test linux-inotify.c code, which is Linux-specific anyway.

saghul commented:
@ahippo Since the fix is just for Linux, I'd be happy with a Linux-only-fsevents-only test. We can add a more generic test with your QUEUE improvements.

ahippo commented:
> @ahippo Since the fix is just for Linux, I'd be happy with a Linux-only-fsevents-only test. We can add a more generic test with your QUEUE improvements.

Cool.
I've updated the pull request with the same test returning RETURN_SKIP() for non-Linux platforms.

saghul commented:
CI: https://ci.nodejs.org/job/libuv+any-pr+multi/188/

saghul commented:
Things look good. I'd like to have the test be done properly though. If you have no time now, I propose the following (I will make the changes if you agree):

- have a test which tests that prepare / idle / check work fine (all platforms)
- have a test which just checks for fsevents on Linux, since that's what this issue is fixing

I can amend your patches if you don't have the time and are ok with that.

ahippo commented:
> Things look good. I'd like to have the test be done properly though. If you have no time now, I propose the following (I will make the changes if you agree):
> 
> have a test which tests that prepare / idle / check work fine (all platforms)
> have a test which just checks for fsevents on Linux, since that's what this issue is fixing
> I can amend your patches if you don't have the time and are ok with that.

Of course, you can go ahead and make whatever changes you want.
Or there are still a few chances that I'll have time this Friday to do that myself.

saghul commented:
I won't have time to work on those until next Tuesday, so we can see then.
On Nov 25, 2015 07:38, "Andrey Mazo" <notifications@github.com> wrote:

> Things look good. I'd like to have the test be done properly though. If
> you have no time now, I propose the following (I will make the changes if
> you agree):
>
> have a test which tests that prepare / idle / check work fine (all
> platforms)
> have a test which just checks for fsevents on Linux, since that's what
> this issue is fixing
> I can amend your patches if you don't have the time and are ok with that.
>
> Of course, you can go ahead and make whatever changes you want.
> Or there are still a few chances that I'll have time this Friday to do
> that myself.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/libuv/libuv/pull/621#issuecomment-159511785>.
>

saghul commented:
Thanks a lot Andrey! Landed with fixes in 46343764d5aa4af0a2e939b6f0f1fad798af43ef and 6060841edca2393345b119c93d11b102b6ce650f

ahippo commented:
> Thanks a lot Andrey! Landed with fixes in 4634376 and 6060841

@saghul, sorry for the late reply.
Thank you very much for taking care of this and improving the test!

I'll rebase other commits in #620 then.
